### PR TITLE
Use immutable support methods

### PR DESCRIPTION
**Pull request by IzumiSy**
I really appreciated the implementation of Immutable.js support you have done, but that had a little awkward implementation which did not use methods offered by Immutable.js itself and it caused an error. This PR makes use of them instead of original one for fixing errors.

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -7,22 +8,11 @@ import * as React from 'react'
  */
 export function equals<T>(a: T, b: T): boolean {
   if (isImmutable(a) && isImmutable(b)) {
-    return a.is(b)
```
Comment (IzumiSy) [inline src/utils.ts]: I got an error from here, because is` method is not an member of List class.

**Comment by bcherny** [comment]
Hi @izumySy! I appreciate the bug report, but the code in your PR will fail for any users that are not using Immutable. Can you give me more details about the error (or better yet, a repro case)? List should have .is() defined on it - is there some other way to use List where that method won’t be defined?

**Comment by IzumiSy** [comment]
At first, I found that with your implementation, I get this error.

selection_049

This is because `is` method is not an member of `List` class, and it is just an static method accepting two Immutable objects as arguments. I tried to fix the error with a correct usage of it, but now I found that my PR fails. I am now trying to manage it. Sorry for that!

**Comment by IzumiSy** [comment]
See implementation of `Immutable#is`. This at first uses triple equality. I think `equals` method might be replaced with it.

https://github.com/facebook/immutable-js/blob/master/src/is.js

**Comment by bcherny** [comment]
Again, this pulls in the entirety of immutable.js as a dependency, and puts the burden on the consumer to use a tree shaking optimizer to remove it. We don't want to do that.

Please:

1. Remove the import from `immutable`
2. Copy the whole `.is` function that you linked into react.ts, and rename is to `equals`
3. Where the function calls `isValueObject`, instead call the `isImmutable` function that I cribbed from Immutable

**Comment by bcherny** [comment]
Alternatively, install [immutable-is](https://www.npmjs.com/package/immutable-is) as a dependency and use that instead of copy+pasting code. Whatever you prefer!

**Comment by bcherny** [comment]
And a third option: undo all of your changes, and just update 1 line in my code:

```diff
- return a.is(b)
+ return a.equals(b)
```

**Comment by IzumiSy** [comment]
> this pulls in the entirety of immutable.js as a dependency, and puts the burden on the consumer to use a tree shaking optimizer to remove it. We don't want to do that.

I see. Got the point. I will remove that from deps.

> Alternatively, install immutable-is as a dependency and use that instead of copy+pasting code. Whatever you prefer!

I think this might be the best and neatest option for us. I will take this up.

**Comment by IzumiSy** [comment]
Now I removed immutable.js from deps and added immutable-is 💪

**Comment by bcherny** [comment]
Nice work @IzumiSy! Can you also do what I should have done, and add test for a few immutable collections (or all of them!) to make sure that they work correctly with the new behavior? Ie. Create a store with Immutable properties, update them, make sure that React caused the DOM to update.

**Comment by IzumiSy** [comment]
Done. You can see if my test works or not by switching `is(a, b)` around `a === b`.

File: src/utils.ts (hunk 1 above)
Comment (bcherny) [reply src/utils.ts]: Question - if we just changed this line to `.equals` instead of `.is`, would that fix your issue?

Comment (IzumiSy) [reply src/utils.ts]: Looks yes. If so, we don't have to have `immutable-is` anymore and it would be better.

Comment (bcherny) [reply src/utils.ts]: Let's do that, and keep the test you wrote.

File: src/utils.ts [hunk 2]
```diff
@@ -7,19 +8,15 @@ import * as React from 'react'
...
 }
 
-export type Immutable = {
-  is(b: any): boolean
-}
-
-export function isImmutable(a: any): a is Immutable {
-  return !!a && typeof a === 'object' && (
-    '@@__IMMUTABLE_ITERABLE__@@' in a
-    || '@@__IMMUTABLE_RECORD__@@' in a
```
Comment (bcherny) [inline src/utils.ts]: Please keep this as-is. `a` is an `any`, not an `object`.

File: test/immutable.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        <div>{fruits.get('banana')}</div>
+      </div>
+    );
+  });
+
+  withElement(TestingComponent, _ => {
+    Simulate.click(_.querySelector('button')!)
+    Simulate.click(_.querySelector('button')!)
+    Simulate.click(_.querySelector('button')!)
+    t.is(renderCount, 1);
```
Comment (bcherny) [inline test/immutable.tsx]: Nice!

File: src/utils.ts (hunk 2 above)
Comment (IzumiSy) [reply src/utils.ts]: OK!

**Comment by IzumiSy** [comment]
Finally we got the cleanest changes. Thanks!

> And a third option: undo all of your changes, and just update 1 line in my code

I missed this comment 🙏

**Comment by bcherny** [comment]
Published 3.2.1

**Pull request merged**